src/KeyRegistry.sol

```diff
@@ -0,0 +1,325 @@
+        _transferOwnership(_owner);
+
+        gracePeriod = _gracePeriod;
+        idRegistry = IdRegistry(_idRegistry);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               MODIFIERS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Revert if caller does not own the associated fid.
+     */
+    modifier onlyFidOwner(uint256 fid) {
+        if (idRegistry.idOf(msg.sender) != fid) revert Unauthorized();
+        _;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              VIEWS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Retrieve signer state for a fid/scope/key tuple.
+     *
+     * @param fid   The fid associated with the key.
+     * @param scope The key's numeric scope. In the initial migration all keys will have scope 1.
+     * @param key   Bytes of the signer's public key.
+     *
+     * @return Signer struct including state and message merkle root for frozen signers.
+     */
+    function signerOf(uint256 fid, uint256 scope, bytes calldata key) external view returns (Signer memory) {
+        return signers[fid][scope][key];
+    }
+
+    /**
+     * @notice Check if the contract has been migrated.
+     *
+     * @return true if the contract has been migrated, false otherwise.
+     */
+    function isMigrated() public view returns (bool) {
+        return signersMigratedAt != 0;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                    EXTERNAL KEY REGISTRATION
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Register a public key to a fid/scope pair, setting the signer state to AUTHORIZED.
+     *
+     * @param fid   The fid associated with the key. Caller must own the provided fid.
+     * @param scope The key's numeric scope.
+     * @param key   Bytes of the signer's public key to authorize.
+     */
+    function register(uint256 fid, uint256 scope, bytes calldata key) external onlyFidOwner(fid) {
+        _register(fid, scope, key);
+    }
+
+    /**
+     * @notice Revoke a public key associated with a fid/scope pair, setting the signer state to REVOKED.
+     *         The key must be in the AUTHORIZED or FROZEN state.
+     *
+     * @param fid   The fid associated with the key. Caller must own the provided fid.
+     * @param scope The key's numeric scope.
+     * @param key   Bytes of the signer's public key to revoke.
+     */
+    function revoke(uint256 fid, uint256 scope, bytes calldata key) external onlyFidOwner(fid) {
+        Signer storage signer = signers[fid][scope][key];
+        if (signer.state != SignerState.AUTHORIZED && signer.state != SignerState.FROZEN) revert InvalidState();
+
+        signer.state = SignerState.REVOKED;
+        emit Revoke(fid, scope, key, key);
```

Comment:
after thinking through the "freezing" logic, i think its a little too complicated to enforce right now on the hubs, and we should just get rid of the state to simplify this contract